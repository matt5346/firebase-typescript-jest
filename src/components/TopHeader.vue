<template>
  <div class="top-header">
    <div class="top-header__inner">
      <h1>User status</h1>
      <h2>
        <span>Logged in </span>
        <span v-if="loggedIn" class="status"><b>Yes</b></span>
        <span v-else class="status"><b>No</b></span>
      </h2>
    </div>
    <div>
      <button @click="signOut" v-if="loggedIn">Logout</button>
    </div>
  </div>
</template>

<script lang="ts">
import Component from 'vue-class-component';
import { Vue } from 'vue-property-decorator';
import { signOut, getAuth, onAuthStateChanged } from 'firebase/auth';
import app from '../firebase';

@Component
export default class TopHeader extends Vue {
  loggedIn = false;

  auth = getAuth(app);

  mounted() {
    this.setupFirebase();
  }

  setupFirebase():void {
    onAuthStateChanged(this.auth, (user) => {
      console.log(user, 'user onAuthStateChanged');
      this.loggedIn = !!user;
    });
  }

  async signOut():Promise<void> {
    try {
      const data = await signOut(this.auth);
      this.loggedIn = false;
      console.log(data, 'data');
      this.$router.replace({ name: 'loginView' });
    } catch (err) {
      console.log(err, 'err signOut');
    }
  }
}
</script>

<style lang="scss">
.top-header {
  position: absolute;
  display: flex;
  justify-content: space-between;
  align-items: center;
  top: 20px;
  left: 0;
  width: 100%;
  padding: 20px 30px;
}

.top-header__inner {
  display: flex;
  flex-direction: column;
  justify-content: flex-start;

  span {
    font-size: 20px;
    &.status {
      color: #ff4b2b;
      text-decoration: underline;
    }
  }
}
</style>
