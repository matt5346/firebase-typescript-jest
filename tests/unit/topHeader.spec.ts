import Vue from 'vue';
import { shallowMount, Wrapper } from '@vue/test-utils';
import { CombinedVueInstance } from 'vue/types/vue';
import TopHeader from '@/components/TopHeader.vue';

describe('topHeader.vue', () => {
  let wrapper: Wrapper<TopHeader, Element>;
  let cmp;
  let vm: CombinedVueInstance<Record<never, any> & Vue, object, object, object, Record<never, any>>;

  beforeEach(() => {
    cmp = Vue.extend(TopHeader);
    vm = new cmp().$mount();
    wrapper = shallowMount(TopHeader, {
      // methods: { setupFirebase: () => {} },
    });
  });

  it('renders', () => {
    expect(wrapper.exists()).toBe(true);
  });

  it('has the expected html structure', () => {
    expect(vm.$el).toMatchSnapshot();
  });

  it('checking h1 text in top header', () => {
    expect(wrapper.find('h1').text()).toBe('User status');
  });
});
