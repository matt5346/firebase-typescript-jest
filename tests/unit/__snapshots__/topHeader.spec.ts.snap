// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`topHeader.vue has the expected html structure 1`] = `
<div
  class="top-header"
>
  <div
    class="top-header__inner"
  >
    <h1>
      User status
    </h1>
     
    <h2>
      <span>
        Logged in 
      </span>
       
      <span
        class="status"
      >
        <b>
          No
        </b>
      </span>
    </h2>
  </div>
   
  <div>
    <!---->
  </div>
</div>
`;
